javaslang/src/main/java/javaslang/collection/Iterator.java

```diff
@@ -658,36 +622,45 @@ public Double getNext() {
         } else if (step == 0) {
             throw new IllegalArgumentException("step cannot be 0");
         } else if (from == toInclusive) {
-            return Iterator.of(from);
-        } else if (step * (from - toInclusive) > 0) {
-            return Iterator.empty();
+            return of(from);
+        } else if (Math.signum(step) == Math.signum(from - toInclusive)) {
+            return empty();
         } else {
-            return new AbstractIterator<Double>() {
-
-                double prev = Double.NaN;
-                double curr = from;
-                boolean hasNext = true;
+            if (step > 0) {
+                return new AbstractIterator<Double>() {
+                    final double end = Math.nextUp(toInclusive);
+                    double i = from;
+
+                    @Override
+                    public boolean hasNext() {
+                        return i < end;
+                    }
 
-                @Override
-                public boolean hasNext() {
-                    return hasNext;
-                }
+                    @Override
+                    public Double getNext() {
+                        final double next = i;
+                        i = Math.max(i + step, Math.nextUp(i));
```

Comment:
Maybe, but it is a problem when using floating points. We need an appropriate solution. At $work team members implemented a calculation core for financial products and there were error cases because of the step problem.